api/environments/sdk/serializers.py

```diff
@@ -46,9 +46,22 @@ def create(self, validated_data):
             else STRING,
         }
 
-        return Trait.objects.update_or_create(
-            identity=identity, trait_key=trait_key, defaults=defaults
-        )[0]
+        try:
+            trait_obj = Trait.objects.get(identity=identity, trait_key=trait_key)
+            updated = False
+            for field, value in defaults.items():
+                existing_value = getattr(trait_obj, field)
+                if value is not None and value != existing_value:
+                    setattr(trait_obj, field, value)
+                    updated = True
+            if updated:
+                trait_obj.save()
+        except Trait.DoesNotExist:
+            trait_obj = Trait.objects.create(
+                identity=identity, trait_key=trait_key, **defaults
+            )
```

Comment:
I wonder if this would be better defined on the `TraitManager` class itself? Maybe `update_if_changed_or_create` or something? That way the logic in this serializer stays cleaner, and we can much more easily unit test this logic.

Reply:
Thank you for feedback and suggestions. 

However, currently, my Python skill level might not be sufficient to implement the proposed changes accurately and efficiently. I believe that these modifications would be better suited for someone with deeper knowledge and expertise in it.

I'd be utterly grateful if someone from the team could take ownership of making these suggested improvements, ensuring the best result for the codebase and the project.

Reply:
@cat5e-inv I've actually been looking into this further and I'm not wholly convinced that this change is a good idea. As you can see from the django code [here](https://github.com/django/django/blob/3.2.25/django/db/models/query.py#L596-L614), the lock is implemented to ensure that we don't have conflicting writes in the database. If we remove that lock, we could end up with inconsistent behaviour. For example: 

Request 1 comes in with an update to trait `trait_key=foo` to set `trait_value=bar`, retrieves the trait from the database, and starts trying to determine whether it needs to update or not
Request 2 comes in with an update to the same trait, to set `trait_value=baz`, retrieves the trait from the database and (quicker than request 1) updates the `trait_value` successfully. 
Request 1 finishes determining whether it needs to update the trait and sets the `trait_value` to `bar`. 

Since we can't guarantee the order of execution in this scenario (because threading), we would end up with the final result being that the trait takes the value of request 1, when it should have taken the value of request 2. 

I can certainly understand where you're coming from here, and the actual scenario where an issue like this might occur is pretty slim, but I am certainly reluctant to implement this change based on this. We could perhaps implement an environment variable with a warning in the docs that it might result in inconsistent writes, but will likely reduce DB lock time.

Reply:
Do you have any quantifiable data around these locks causing a problem for you?

Reply:
@matthewelwell @cat5e-inv the screenshot in https://github.com/Flagsmith/flagsmith/issues/3595 shows the Cloud SQL PG instance metrics, and how locks and update volume drops.

The locking issue in particular is probably not going to be a big deal for most users, because one needs to concurrently update the same traits of the same identity. Even in our case, I must admit, that's a deficiency of our approach, not Flagsmith's.

If we concentrate just on the locking, however, and not on the volume of writes, what do you think about the use of PG's conflict handling? If each trait update writes anyway, an `INSERT...ON CONFLICT UPDATE` at least removes the need to run a `SELECT...FOR UPDATE`. The only problem is that it doesn't work with standard `update_or_create()`: https://django-postgres-extra.readthedocs.io/en/master/conflict_handling.html. This approach should also be slightly more performant due to at least one less roundtrip. This code would not be database-agnostic, however, and I'm not sure how much of a problem that is.

The originally proposed change, however, reduces the sheer volume of updates for the situation where trait values don't change that often. That seems to be a good thing. However, you're right that it breaks the ordering of updates, and allows the oldest transaction to "win". I tried to think of a way to avoid the current pessimistic locking approach but failed to come to something concrete that is also not overly complicated. I'll write out my thoughts anyway, perhaps someone will get inspired.

One way out could be to use optimistic locking, but that would necessitate using some way of tracking not only the record version, but also transaction itself against its peers. That should be possible with PG's xmin, but at the very least will be quite complicated. Another option is to run this particular transaction under SERIALIZABLE isolation in PG, which will reduce locking substantially, but would also cause concurrent transactions to fail on commit due to serialization failures. First committer wins with PG SSI (https://wiki.postgresql.org/wiki/SSI), and I think that would still require the same tracking to decide whether to re-try or not. If t1 commits when t2 and t3 have just finished their reads, both t2 and t3 will fail, and the earlier transaction wins. If t2 and t3 retry simultaneously, it's not possible to say whether the end result will be that of t2 or t3 unless they can order themselves.

Reply:
> This code would not be database-agnostic, however, and I'm not sure how much of a problem that is.

Yes, this is unfortunately an issue for us as we support multiple database engines for our Enterprise offering. We could add some conditional logic to separate out based on database engine, but it gets quite complicated and I'd rather avoid it if possible. 

Could you compare the traits on your side before sending the update using the `getTrait` or `getTraits` method in the Kotlin client?

Reply:
@arronax @cat5e-inv do you have anything further to add here?

Reply:
@cat5e-inv thanks for the contribution to the kotlin client and apologies for the delay in getting back to you. 

As mentioned on the PR, I have merged it (too hastily), but I cannot release a new version as the implementation is not compatible with our SaaS platform. I have a PR [here](https://github.com/Flagsmith/flagsmith-kotlin-android-client/pull/47) to revert the change, but if you're able to submit a small update to use the endpoint I mentioned on the comment on the PR then that would likely be the best next step.